### PR TITLE
Upgrade test suite to TS

### PR DESCRIPTION
**Pull request by ryanio**
This PR:

1. Upgrades test suite to TypeScript
1. Adds new types `BufferCallback` and `ProveCallback` to satisfy implicit type errors in tests
1. Removes use of deprecated `new Buffer()`
1. Adds convenience script `npm run benchmarks` for previously existing `benchmarks` folder

### CODE REVIEW

**Comment by ryanio** [comment]
I added the file renames as a separate commit to try to encourage GitHub to show the files as renamed instead of all deleted and all inserted, but I guess it didn't work for some files 🤷🏻‍♂️sorry about that, makes the diff really challenging to compare. Let me know if you have any tips or suggestions.

File: benchmarks/checkpointing.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    })
+  },
+  function () {
+    console.info(
+      'Average Execution time (hr): %ds %dms',
+      avg[0] / samples,
+      avg[1] / 1000000 / samples,
+    )
+  },
+)
```
Comment (holgerd77) [inline benchmarks/checkpointing.ts]: I've no idea about these benchmarking files TBH, did you test this and are they working (now)?

If they do and if they do some useful things, can you give this an additional entry to the README?

File: package.json [hunk 2]
```diff
@@ -48,6 +49,7 @@
   ],
   "license": "MPL-2.0",
   "dependencies": {
+    "@types/async": "^3.0.8",
     "async": "^2.6.1",
```
Comment (holgerd77) [inline package.json]: We want at some point getting rid of the `async` dependencies - just as a side note - see issue here https://github.com/ethereumjs/organization/issues/55

File: package.json [hunk 3]
```diff
@@ -22,7 +23,7 @@
     "tsc": "ethereumjs-config-tsc",
     "test": "npm run test:node && npm run test:browser",
     "test:browser": "npm run build && karma start karma.conf.js",
-    "test:node": "npm run build && tape test/*.js | tap-prettify -"
+    "test:node": "npm run build && tape -r ts-node/register test/*.ts | tap-prettify -"
```
Comment (holgerd77) [inline package.json]: Just gave the tests a manual local run, looks good, test number is preserved (127 tests). 😋

File: package.json [hunk 4]
```diff
@@ -8,6 +8,7 @@
     "dist/**/*"
   ],
   "scripts": {
+    "benchmarks": "npm run build && ts-node benchmarks/index.ts",
```
Comment (holgerd77) [inline package.json]: Ah, didn't see this when doing the comment on the benchmark files above.

However this fails for me with a completely fresh node module reinstall with several implicit `any` type errors. Any comment on that?

File: src/baseTrie.ts [hunk 5]
```diff
@@ -62,12 +62,12 @@ export class Trie {
       }
     }
 
-    proofTrie.db.batch(opStack, (e?: Error) => {
-      cb(e, proofTrie)
+    proofTrie.db.batch(opStack, (err) => {
+      cb(err, proofTrie)
```
Comment (holgerd77) [inline src/baseTrie.ts]: Not really getting this change here, could you elaborate?

File: benchmarks/checkpointing.ts (hunk 1 above)
Comment (ryanio) [reply benchmarks/checkpointing.ts]: Yes they work for me, the files were already present (I only upgraded them to TS). I didn't look into how useful they are yet but it's on my task list to investigate after some promisify work :)

File: package.json (hunk 2 above)
Comment (ryanio) [reply package.json]: ah ok thanks

File: package.json (hunk 4 above)
Comment (ryanio) [reply package.json]: Thanks, I just did a fresh install and ran into the errors, will look at fixing.

File: src/baseTrie.ts (hunk 5 above)
Comment (ryanio) [reply src/baseTrie.ts]: yes I just renamed `e` to `err` and removed the explicitly defined type (already typed as [`ErrorCallback`](https://github.com/ethereumjs/merkle-patricia-tree/blob/master/src/db.ts#L85))

Comment (holgerd77) [reply src/baseTrie.ts]: Ah ok, thanks!

File: package.json (hunk 4 above)
Comment (holgerd77) [reply package.json]: Can confirm that this works now.

File: benchmarks/random.js [hunk 6]
```diff
@@ -1,23 +1,23 @@
 // https://github.com/ethereum/wiki/wiki/Benchmarks
 'use strict'
-const Trie = require('../')
-const ethUtil = require('ethereumjs-util')
-const async = require('async')
+import * as async from 'async'
+import * as ethUtil from 'ethereumjs-util'
+const Trie = require('../dist/index.js').BaseTrie
```
Comment (holgerd77) [inline benchmarks/random.js]: Just as a note: benchmarks now moved to be run on `dist`.

File: package.json (hunk 2 above)
Comment (ryanio) [reply package.json]: Oops this should in devDeps, I’ll move it

Comment (holgerd77) [reply package.json]: Oh, you are right. Already merged, just do a short follow-up PR eventually.

Comment (ryanio) [reply package.json]: No prob, will include in next PR for `promisify` work, hopefully today or tomorrow.

**Pull request merged**